Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/2-types/4-tuples.md

# Tuples

A tuple is a fixed-size, ordered collection of values, where each element can have a different type. Tuples provide a way to group values by position, without the overhead of defining a record of named fields.

A tuple is grouped together in parentheses (`value1`, `value2`, `value3`). The type of a tuple is determined by types of its elements, such as ([`Text`](https://internetcomputer.org/docs/motoko/base/Text), [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat), [`Bool`](https://internetcomputer.org/docs/motoko/base/Bool)). The values inside a tuple are evaluated in order from left to right. Tuples are immutable and their components cannot be modified after creation (unlike record fields, which can be declared mutable).  

> A tuple with zero elements is called the **unit value**, written as `()`. Since it carries no other data, it represents a trivial or token value.  
> It's type is the **unit type**, also written `()`.  

Reviewer: ```suggestion
A tuple with zero elements is called the **unit value**, written as `()`. Since it carries no other data, it represents a trivial or token value. It's type is the **unit type**, also written `()`.  
```

A tuple is grouped together in parentheses (`value1`, `value2`, `value3`). The type of a tuple is determined by types of its elements, such as ([`Text`](https://internetcomputer.org/docs/motoko/base/Text), [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat), [`Bool`](https://internetcomputer.org/docs/motoko/base/Bool)). The values inside a tuple are evaluated in order from left to right. Tuples are immutable and their components cannot be modified after creation (unlike record fields, which can be declared mutable).  

> A tuple with zero elements is called the **unit value**, written as `()`. Since it carries no other data, it represents a trivial  or token value.  

Reviewer: ```suggestion
A tuple with zero elements is called the **unit value**, written as `()`. Since it carries no other data, it represents a trivial or token value.  
```

> A tuple with zero elements is called the **unit value**, written as `()`. Since it carries no other data, it represents a trivial  or token value.  
> It's type is the **unit type**, also written `()`.  
> 
> Unit values are typically used as dummy argument or return values for functions that don't require arguments or don't produce results. However, they may still produce side effects like printing to the console.
> 

Reviewer: ```suggestion
A tuple with zero elements is called the **unit value**, written as `()`. Since it carries no data, it represents a trivial or token value. Its type is the **unit type**, also written `()`.

Unit values are typically used as placeholder arguments or return values for functions that don't require input or don't return meaningful results. However, such functions may still produce side effects, such as printing to the console or modifying state.
```

A tuple with zero elements is called the **unit value**, written as `()`. It represents a trivial result or token return value.  
It's type is the **unit type**, also written `()`.  

> Unit values are typically used as dummy argument or return values for functions that don't require arguments, or don't produce results (but may produce side-effects eg: printing to the console).

Reviewer: ```suggestion
Unit values are typically used as dummy argument or return values for functions that don't require arguments or don't produce results. However, they may still produce side effects like printing to the console.
```

Tuples are useful for returning multiple values from a function without requiring a separate data structure.

```motoko
func getUserInfo() : (Text, Nat) {
    ("Ghost", 30);
};

getUserInfo();
```

> ## Nesting tuples

Reviewer 2: I would move the section `## Nesting tuples` to the end, since it relies on pattern matching of tuples.

> Tuples can be used to represent coordinate pairs, allowing for structured calculations such as finding the gradient of a line. The gradient (or slope) between two points $(x_1, y_1)$ and $(x_2, y_2)$ is calculated using the formula:

Reviewer 2: I think the following example might be clearer with a couple of type definitions:

```
//  Point is a tuple
type Point = (Float, Float);
// Line is a tuple of points, that is, a nested tuple
type Line = (Point, Point);
func calculateGradient(line: Line) : ?Float {
let ((x1, y1), (x2, y2)) = line;  
  if (x1 == x2) {  
    null // Gradient is undefined for a vertical line  
  } else {  
    ?((y2 - y1) / (x2 - x1)) // Wraps the result as an option  
  }  
}  
let line : Line = ((2.0, 3.0), (5.0, 7.0));
calculateGradient(line); // should show the gradient when run
```

> Tuples can be used to represent coordinate pairs, allowing for structured calculations such as finding the gradient of a line. The gradient (or slope) between two points $(x_1, y_1)$ and $(x_2, y_2)$ is calculated using the formula:

Reviewer 2: I would move ## Nesting tuples to the end, since it relies on pattern matching of tuples which is only introduce later.

let users : [(Text, Nat)] = [("Motoko", 25), ("Ghost", 30)];
```

This structure efficiently represents a collection of key-value pairs without requiring a dedicated [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) type.

## Pattern matching on tuples

> As well as the dot notation, tuple can be decomposed using tuple patterns. Combined with `let` or `switch`, this let you access the components of a tuple by simple pattern matching.
> 

Reviewer: ```suggestion
In addition to dot notation, tuples can be decomposed using tuple patterns. When combined with `let` or `switch`, this allows you to access the components of a tuple through simple pattern matching.
```

```motoko no-repl
let users : [(Text, Nat)] = [("Motoko", 25), ("Ghost", 30)];  

let (firstUserName, _) = users[0] // "Motoko" 
```

> The array `users` contains tuples, where each tuple represents a user with a [`Text`](https://internetcomputer.org/docs/motoko/base/Text) name and a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) age. `users[0]` retrieves the first tuple in the array. The `let` pattern extracts only the name while ignoring the second element using the wildcard pattern (`_`).

Reviewer: ```suggestion
The array `users` contains tuples, where each tuple represents a user with a [`Text`](https://internetcomputer.org/docs/motoko/base/Text) name and a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) age.  
Accessing `users[0]` retrieves the first tuple in the array.  
A `let` pattern can then be used to extract just the name, ignoring the age by using the wildcard pattern (`_`).
```

## Nesting tuples

Tuples can be used to represent coordinate pairs, allowing for structured calculations such as finding the gradient of a line. The gradient (or slope) between two points $(x_1, y_1)$ and $(x_2, y_2)$ is calculated using the formula:

$$
m = \frac{y_2 - y_1}{x_2 - x_1}
$$

Using nested tuples, this can be implemented in Motoko as follows:

```motoko

> type Point = (Float, Float);
> // Line is a tuple of points, that is, a nested tuple
> type Line = (Point, Point);

Reviewer 2: ```suggestion
  type Point = (Float, Float);
  // Line is a tuple of points, that is, a nested tuple
  type Line = (Point, Point);
```
The indentation is inconsistent. Either indent the first three lines (as suggested) or undent the remaining ones.

## Pattern matching on tuples

In addition to dot notation, tuples can be decomposed using tuple patterns. When combined with `let` or `switch`, this allows you to access the components of a tuple through simple pattern matching.
```motoko no-repl
let users : [(Text, Nat)] = [("Motoko", 25), ("Ghost", 30)];  

let (firstUserName, _) = users[0] // "Motoko" 
```

The array `users` contains tuples, where each tuple represents a user with a [`Text`](https://internetcomputer.org/docs/motoko/base/Text) name and a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) age.  
Accessing `users[0]` retrieves the first tuple in the array.  
A `let` pattern can then be used to extract just the name, ignoring the age by using the wildcard pattern (`_`).

## Nesting tuples

Tuples can be used to represent coordinate pairs, allowing for structured calculations such as finding the gradient of a line. The gradient (or slope) between two points $(x_1, y_1)$ and $(x_2, y_2)$ is calculated using the formula:

$$
m = \frac{y_2 - y_1}{x_2 - x_1}
$$

> 

Reviewer: It renders properly in the portal

let users : [(Text, Nat)] = [("Motoko", 25), ("Ghost", 30)];
```

This structure efficiently represents a collection of key-value pairs without requiring a dedicated [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) type.

> ## Extracting values using switch

Reviewer 2: ```suggestion
## Decomposing tuples using tuple patterns
```
